Add l1 verifier

joshklop opened the pull request:
Key additions and modifications:
- Add the `L1HeadKey` key to the database. It holds the latest `core.L1Head`
- Add `blockchain.SetL1Head`
- The `l1` package contains most of the new logic.
- We can mock the `l1.Subscriber` interface in tests. In practice, the interface is implemented by the `l1.StarknetFilterer`, which was generated using go-ethereum's abigen tool: `abigen --abi starknetAbi.json --pkg starknet`. Since we're only using the filterer, we can remove a lot of the generated code if we want.
  - codecov should ignore the generated `l1/contract/starknet.go` file.  

`node.Run` -> `l1.Client.Run` -> (update is received from subscription) -> `blockchain.SetL1Head`

omerfirmak commented on l1/l1.go:
```diff
@@ -0,0 +1,70 @@
+			return err
+		case <-ctx.Done():
+			return nil
+		case update := <-updateChan:
```
we should check for update.Raw.Removed

https://github.com/ethereum/go-ethereum/blob/master/core/types/log.go#L53-L55

omerfirmak commented:
I think we should implement a "confirmation period" like approach and only update the L1 head when we are pretty confident that the log will not be removed due to a reorg.

We will have to subscribe to both logs and new blocks. 

- On new logs; 
 If `Removed` is false, save the log and the height that the log was emitted on, in a FIFO queue.
 If `Removed` is true, find that log in the queue and remove

- On new blocks;
See if the oldest log in FIFO queue is atleast `N` blocks old. 
   If yes, pop the log from queue and update the L1 head.
   If no, do nothing.

joshklop commented on l1/l1.go:
```diff
@@ -0,0 +1,134 @@
+			return nil
+		case header := <-headerChan:
+		Outer:
+			for {
```
I think the loop should terminate (see `break Outer` in the `default` case). I added a test and it passes, but maybe I'm missing something?

I agree the order of messages doesn't matter in practice. However, it does allow us to make the tests deterministic. That's why I ended up nesting the `select`s. 

I think waiting 2 epochs for L1 is maybe a bit much? The overwhelming majority of [past reorgs](https://etherscan.io/blocks_forked?p=1) appear to be single-block reorgs. I couldn't find a non-single block reorg in the last 54 days. What do you think?

omerfirmak replied:
> I agree the order of messages doesn't matter in practice. However, it does allow us to make the tests deterministic. That's why I ended up nesting the selects.

I would avoid doing that just for the sake of easier testing. Could we have updated the test instead?

joshklop replied:
> doesnt that just jump to start of the loop again?

It should break out of the `Outer` loop. [This test](https://github.com/NethermindEth/juno/blob/37a5d711173450f0cc3bb1f2e7f348d1217a548a/l1/l1_test.go#L122) wouldn't succeed if it didn't behave as expected.

> Could we have updated the test instead?

I don't think so. I feel the tests should be agnostic to the order the messages are delivered. With the nested selects, we only rely on the assumption: "if we receive logs, then we also receive a header around the same time."

omerfirmak replied:
> we only rely on the assumption: "if we receive logs, then we also receive a header around the same time."

Also we rely on events coming before the new header, dont we? Nested select is a bit racey, although since headers are pretty frequent it might not make a difference in the end.

Assume you get a header for height `N`, but events for that height `N` arrive after some time. So nested select will exit thru the default case since the events are not in the channels yet and we will only be processing the events after the next header for `N+1`.

omerfirmak replied:
Not really. Other than making the flat-select work

joshklop replied:
The flat-select has the same issue, right?

omerfirmak replied:
Maybe, I might have not really thought this thru. We dont really need to fix this, it is not an issue at all. It is just an implicit `+1` to the confirmation period.

omerfirmak commented on node/node.go:
```diff
@@ -145,7 +149,24 @@ func (n *Node) Run(ctx context.Context) {
 	http := makeHTTP(n.cfg.RPCPort, rpc.New(n.blockchain, n.cfg.Network), n.log)
 
-	n.services = []service.Service{synchronizer, http}
+	ethClient, err := ethclient.Dial(n.cfg.EthNode)
+	if err != nil {
+		n.log.Errorw("Cannot connect to Ethereum node", "err", err)
+		return
+	}
+	coreContractAddress, err := n.cfg.Network.CoreContractAddress()
+	if err != nil {
+		n.log.Errorw("Error finding core contract address for network", "err", err, "network", n.cfg.Network)
+		return
+	}
+	coreContractFilterer, err := contract.NewStarknetFilterer(*coreContractAddress, ethClient)
+	if err != nil {
+		n.log.Errorw("Failed to set up filterer for L1 contract", "err", err, "network", n.cfg.Network, "ethereumNode", n.cfg.EthNode)
+		return
+	}
+	l1Client := l1.NewClient(l1.NewWatcher(ethClient, coreContractFilterer), n.blockchain, n.log)
```
I think we can move all this to l1.go
It might make testing a little harder but we can always introduce a helper to make the dependency injection work.

omerfirmak commented on l1/l1.go:
```diff
@@ -0,0 +1,111 @@
+		if blockNumber > l1BlockNumber-ConfirmationPeriod {
+			continue
+		}
+		if err := c.l2Chain.SetL1Update(update); err != nil {
```
we wont be calling `SetL1Update` with ordered `blockNumber`s now.
`SetL1Update` should make sure the incoming update is more recent.

omerfirmak replied:
Dont we still have the same issue here?

when we `range c.q`, we will loop over the `blockNumber`s in random order

wojciechos commented:
In order to make our smoke test work with the added L1 verifier, we need to provide an Ethereum node address. I'm considering whether we should use Infura/Alchemy, run our own node, or ask the core team to provide us with one.

joshklop commented:
- Fixed some issues raised by @omerfirmak 
- Rebased
- Removed unused code that was generated by `abigen`.
- Moved the generated code to `l1/internal`.

wojciechos commented:
I got an error while trying to run Juno with an eth-node via HTTPS(it works for other node implementations). It panicked with this error:

`ERROR node/node.go:205 Service error {"name": "*l1.Client", "err": "subscribe to L1 headers: notifications not supported"}`

When using wss, everything works perfectly. Just wondering, is HTTPS not supported on purpose, or is this something we'll look into?

wojciechos commented:
I guess we should also update README file:

- Run with Docker part - we should add `--eth-node <YOUR-ETH-NODE>` to the command and maybe some short note like: "You should replace `<YOUR-ETH-NODE>` with your actual Ethereum node address. For example, if you're using Infura, your Ethereum node address might look something like this: wss://mainnet.infura.io/ws/v3/your-infura-project-id."
- Mark L1 checkbox in Phase 2

joshklop commented:
> I guess we should also update README file

Done, thanks!

joshklop commented:
A things that can be done in this PR or a follow-up:

- In order to test cases involving multiple block headers, we need to replace the subscriber (at least using the current mocking techniques--we can probably use custom mocks and avoid this issue). The only way to do this without manipulating the public API is by putting the tests in the same package as the code.
- More graceful reorg handling (right now we only update the confirmation queue and assume more logs will come in the future--the local L1 head remains the same).
- Handling failed subscriptions.

kirugan commented on .github/workflows/smoke-tests.yml:
```diff
@@ -40,4 +40,4 @@ jobs:
           TARGET_BLOCK: 100
           TIMEOUT: 5m
-          JUNO_PARAMETERS: "--network mainnet"
+          JUNO_PARAMETERS: "--network mainnet --eth-node ${{ secrets.ETH_NODE }}"
```
Do you need setup somewhere `secrets.ETH_NODE` or it's done already?

joshklop replied:
@wojciechos may know.

wojciechos replied:
It's already done.

kirugan commented on blockchain/blockchain_test.go:
```diff
@@ -398,3 +398,26 @@ func TestState(t *testing.T) {
+	}
+
+	for _, head := range heads {
+		t.Run("update L1 head to block "+string(rune(head.BlockNumber)), func(t *testing.T) {
```
I think you wanted to do smth. like `strconv.Itoa(...)`, currently `string(rune(...))` expression returns A letter and some weird rune LTX for block 1 and 2 respectively

kirugan commented on l1/internal/contract/starknet.go:
```diff
@@ -0,0 +1,103 @@
+// Code generated - DO NOT EDIT.
```
how was it generated?

joshklop replied:
Geth's [abigen](https://geth.ethereum.org/docs/tools/abigen) utility. 
> Abigen creates easy-to-use, type-safe Go packages from Ethereum smart contract definitions known as ABIs. This abstracts away a lot of the complexity of handling smart contract deployment and interaction in Go native applications such as encoding and decoding smart contracts into EVM bytecode

```
abigen --abi coreContract_0.11.1.json --pkg starknet > starknet.go
```

kirugan commented on utils/network.go:
```diff
@@ -97,3 +98,22 @@ func (n Network) ChainID() *felt.Felt {
+		return common.Address{}, errors.New("l1 contract is not available on the integration network")
+	default:
+		// Should not happen.
+		panic(ErrUnknownNetwork)
```
Since there is an error in return signature wdyt about unification i.e. returning this error rather than panicking. As far as I understand both ways lead to same result (node will stop).

kirugan commented on node/node.go:
```diff
@@ -177,6 +180,26 @@ func (n *Node) Run(ctx context.Context) {
+			n.log.Errorw("Error creating ethWatcher", "err", err)
+			return
+		}
+		confirmationPeriod := 16
```
`confirmationPeriod` is for number of blocks right? Also since it's hardcoded now maybe make it const (we can either keep it here or move it outside the func

kirugan commented on l1/eth_subscriber.go:
```diff
@@ -0,0 +1,42 @@
+var _ Subscriber = (*EthSubscriber)(nil)
+
+func NewEthSubscriber(ethClientAddress string, coreContractAddress common.Address) (*EthSubscriber, error) {
+	ethClient, err := ethclient.Dial(ethClientAddress)
```
There is a `ethClient.Close()` method which is not used atm. I think we should call it during shutdown, but maybe it's not critical

kirugan commented on l1/queue.go:
```diff
@@ -0,0 +1,63 @@
+}
+
+func (q *queue) Reorg(forkedBlockNumber uint64) {
+	for l1BlockNumber := range q.q {
```
loop over map in Go is randomized, can it lead to some issues here?

joshklop replied:
I don't think the order matters here since we loop over the entire map.

omerfirmak commented on node/node.go:
```diff
@@ -182,6 +186,26 @@ func (n *Node) Run(ctx context.Context) {
+			n.log.Errorw("Error creating ethWatcher", "err", err)
+			return
+		}
+		defer ethSubscriber.Close()
```
we are immediately closing the `ethSubscriber`, without l1.Run can be called. How does this work?

kirugan commented on node/node.go:
```diff
@@ -11,17 +11,20 @@ import (
 const (
-	defaultPprofPort = uint16(9080)
+	defaultPprofPort   = uint16(9080)
+	confirmationPeriod = uint64(16)
```
constants are untyped so we don't need to specify type here

IronGauntlets commented on l1/queue.go:
```diff
@@ -0,0 +1,63 @@
+//
+// Its abnormal API reflects the fact that it is only meant to be used
+// by the L1 verifier.
+type queue struct {
```
This queue doesn't ensure that there are no missing blocks, which can happen if subscriptions fail and new blocks were created during the creation of a new subscription.

joshklop replied:
Right. You think we should add a comment about that somewhere?

IronGauntlets replied:
I realised it doesn't matter if there are missing blocks all we care about is the difference between the block number and the confirmation period.

IronGauntlets commented on l1/l1.go:
```diff
@@ -0,0 +1,141 @@
+					defer subUpdates.Unsubscribe() //nolint:gocritic
+				case logStateUpdate := <-logStateUpdateChan:
+					c.log.Debugw("Received L1 LogStateUpdate",
+						"number", logStateUpdate.BlockNumber,
```
logStateUpdate.BlockNumber is this Starknet block number or Ethereum Block number?

joshklop replied:
Starknet

IronGauntlets commented on l1/l1.go:
```diff
@@ -0,0 +1,141 @@
+						return err
+					}
+					defer subUpdates.Unsubscribe() //nolint:gocritic
+				case logStateUpdate := <-logStateUpdateChan:
```
How can you be sure that the log emitted here belongs to the `l1height` above? 

If logStateUpdate doesn't contain Ethereum Block Number, how can you be sure that looping here only contains events from a single Ethereum block?

IronGauntlets replied:
The reason I am asking is that I fail to see why we are subscribing to the headers when we can't be sure which ethereum block the log belongs to. 

Using the ethereum block header number to determine whether the log won't be removed requires the assumption that we know which block the log was emitted from. Since we don't have that information making any statement about ethereum block confirmation is superfluous.

I will create separate issue/s for my concerns. For now, we can merge it.

omerfirmak replied:
Each event comes with the block number & hash it is associated with https://github.com/NethermindEth/juno/pull/699/files/9d0eb5e9f39a7ffe787ec242ba588e6b50a2b6d3#diff-7f7fd9a26e7cb24ff29bbaa5130bd597f6ec2c15bec9414dfec4863850ca8570R108-R110

IronGauntlets commented on node/node.go:
```diff
@@ -189,6 +193,26 @@ func (n *Node) Run(ctx context.Context) {
+			n.log.Errorw("Error finding core contract address for network", "err", err, "network", n.cfg.Network)
+			return
+		}
+		var ethSubscriber *l1.EthSubscriber
```
`Subscription` if rename.

joshklop replied:
We need to call EthSubscriber's `Close()` method.

joshklop commented on l1/l1.go:
```diff
@@ -0,0 +1,141 @@
+			return nil
+		case err := <-subHeaders.Err():
+			c.log.Warnw("L1 header subscription failed, resubscribing", "error", err)
+			subHeaders.Unsubscribe()
```
We unsubscribe to release the subscription resources (most notable is the channel).

joshklop commented on node/node.go:
```diff
@@ -189,6 +193,26 @@ func (n *Node) Run(ctx context.Context) {
+			n.log.Errorw("Error creating ethWatcher", "err", err)
+			return
+		}
+		defer ethSubscriber.Close()
```
`EthSubscriber` isn't a service, its an `l1.Subscriber`. That's why it doesn't implement the `Service` interface.

IronGauntlets replied:
You are correct, we do the same thing for closing the db. Apologies for the misunderstanding.

joshklop commented on l1/l1.go:
```diff
@@ -0,0 +1,141 @@
+)
+
+//go:generate mockgen -destination=./mocks/mock_subscriber.go -package=mocks github.com/NethermindEth/juno/l1 Subscriber
+type Subscriber interface {
```
A similar interface is called `Subscriber` in go-ethereum, so I think I'll leave it for now.

joshklop commented on l1/l1.go:
```diff
@@ -0,0 +1,141 @@
+			if err != nil {
+				return err
+			}
+			defer subHeaders.Unsubscribe() //nolint:gocritic
```
> will refer to the same subHeader

We reassign it a few lines above? If I'm missing something, please let me know.

IronGauntlets replied:
Here is an example:

```go
import "fmt"

func main() {
	x := 2
	defer add2(&x)

	x = 0

	// This program will print the following:
	// 2 (not 4, as x was reassigned before it could be evaluated,
        // It will print 4 if add2 accepted a value instead of a reference)
}

func add2(x *int) {
	*x += 2
	fmt.Println(*x)
}
```

Therefore, when the line 68 `Unsubscribe()` method is called it will refer to the subHeader which was most recently assigned, not the initial one. And since you explicitly `Unsubscribe()` if an error occurs. I thought that line 88 was redundant.

IronGauntlets replied:
After reading the documentation, `Unsubscribe()` can be called any number of times, so there won't be a panic.

joshklop commented on l1/l1.go:
```diff
@@ -0,0 +1,141 @@
+		case <-ctx.Done():
+			return nil
+		case err := <-subHeaders.Err():
+			c.log.Warnw("L1 header subscription failed, resubscribing", "error", err)
+			subHeaders.Unsubscribe()
+
+			subHeaders, headerChan, err = c.subscribeToHeaders(subscribeCtx, buffer)
```
I think we can drop lost headers. We should probably pick up lost updates, though. Let me know if you disagree.

joshklop commented on l1/l1.go:
```diff
@@ -0,0 +1,141 @@
+					if err != nil {
+						return err
+					}
+					defer subUpdates.Unsubscribe() //nolint:gocritic
```
Thanks. Conversation continued on other thread.

IronGauntlets commented on core/l1_head.go:
```diff
@@ -0,0 +1,9 @@
+
+import "github.com/NethermindEth/juno/core/felt"
+
+type L1Head struct {
+	BlockNumber uint64
+	BlockHash   *felt.Felt
+	StateRoot   *felt.Felt
+}
```
Yeah, I didn't think about that scenario.

IronGauntlets commented on l1/l1.go:
```diff
@@ -0,0 +1,141 @@
+			return nil
+		case err := <-subHeaders.Err():
+			c.log.Warnw("L1 header subscription failed, resubscribing", "error", err)
+			subHeaders.Unsubscribe()
```
Are you sure that `Unsubscribe()` won't close the `headerChan`?

joshklop replied:
The documentation says `Unsubscribe()` "cancels sending of events, closing the error channel". It should not close the channels it did not create.

Reading the implementation, I can confirm `Unsubscribe()` doesn't close the `headerChan`.

omerfirmak commented on l1/queue.go:
```diff
@@ -0,0 +1,63 @@
+// MaxConfirmed returns the most recent confirmed update and deletes
+// all other confirmed updates.
+func (q *queue) MaxConfirmed(l1Height uint64) *contract.StarknetLogStateUpdate {
+	maxConfirmed := new(contract.StarknetLogStateUpdate)
```
```suggestion
	var maxConfirmed *contract.StarknetLogStateUpdate
```

lets you remove the `if maxConfirmed.BlockNumber == nil` check below

joshklop replied:
Thanks! I'll go with something like this.

This particular change causes a panic in the tests since `maxConfirmed.Raw` is `nil` and we use it a few lines down. Adding a separate `maxConfirmedBlockNumber` var makes it work.

omerfirmak reviewed (approved):
LGTM

joshklop commented on l1/queue.go:
```diff
@@ -0,0 +1,59 @@
+		if l1Height-l1BlockNumber >= q.confirmationPeriod {
+			// We are careful to use >= here. If the max block is zero,
+			// this will still work. Mostly helpful for tests.
+			if l1BlockNumber >= maxConfirmedBlockNumber {
```
Much better. Thanks again!
